File: tests/donchian_test.py
import numpy as np

from taflow import Donchian


def test_donchian_warmup_and_reset() -> None:
    indicator = Donchian(np.array([10.0, 12.0, 11.0]), np.array([8.0, 9.0, 7.0]), 3)
    upper, lower, middle = indicator.compute()
    np.testing.assert_allclose(upper, [np.nan, np.nan, 12.0], equal_nan=True)
    np.testing.assert_allclose(lower, [np.nan, np.nan, 7.0], equal_nan=True)
    np.testing.assert_allclose(middle, [np.nan, np.nan, 9.5], equal_nan=True)
    assert len(indicator) == 3
    assert indicator.reset().value is None

File: crates/taflow-core/src/stream/donchian_test.rs
use super::donchian::Donchian;

#[test]
fn scalar_bulk_and_reset_are_consistent() {
    let high = [10.0, 12.0, 11.0, 13.0];
    let low = [8.0, 9.0, 7.0, 10.0];
    let mut state = Donchian::new(3).unwrap();
    assert_eq!(state.append(high[0], low[0]), None);
    assert_eq!(state.append(high[1], low[1]), None);
    let value = state.append(high[2], low[2]).unwrap();
    assert_eq!(value.upper, 12.0);
    assert_eq!(value.lower, 7.0);
    assert_eq!(value.middle, 9.5);
    state.reset();
    assert_eq!(state.value(), None);
    let mut bulk = Donchian::new(3).unwrap();
    let mut upper = Vec::new();
    let mut lower = Vec::new();
    let mut middle = Vec::new();
    bulk.extend_slices_into(&high, &low, &mut upper, &mut lower, &mut middle)
        .unwrap();
    assert!(upper[0].is_nan() && upper[1].is_nan());
    assert_eq!(upper[2].to_bits(), 12.0f64.to_bits());
    assert_eq!(lower[3].to_bits(), 7.0f64.to_bits());
}
